# Linear and Quadratic Programming Using CVXOPT


The standard form for primal and dual linear programs (slightly different from what was done in class, here RHS of equality and inequality is not kept zero) in cvxopt library is given as follows

$$
\begin{align*}
\text{minimize} \quad &c^T x \\
\text{subject} \quad &Gx \leq h \\
&Ax = b\\
\end{align*}
$$


We would like to solve the following linear programming problem.

$$
\begin{align*}
\text{minimize} \quad &-4x_1 - 5 x_2 \\
\text{subject to} \quad &2x_1 + x_2 \leq 3 \\
&x_1 + 2x_2 \leq 3 \\
&x_1 \geq 0, \: x_2 \geq 0.
\end{align*}
$$

We can write all inequalities as $\leq$ as follows

$$
\begin{align*}
\text{minimize} \quad &-4x_1 - 5 x_2 \\
\text{subject to} \quad &2x_1 + x_2 \leq 3 \\
&x_1 + 2x_2 \leq 3 \\
-&x_1 + 0 x_2 \leq 0 \\
&0 x_1 - x_2 \leq 0.
\end{align*}
$$

To solve this using cvxopt library, we have

$$
G = \begin{bmatrix}
2 & 1 \\
1 & 2 \\
-1 & 0 \\
0 & -1
\end{bmatrix}, \quad c^T = [-4, -5], \quad h = \begin{bmatrix}
3 \\
3 \\
0 \\
0
\end{bmatrix}
$$


In [2]:
! python -m pip install cvxopt

You should consider upgrading via the 'C:\Users\space\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


In [6]:
from cvxopt import matrix, solvers
c = matrix([-4., -5.])
G = matrix([[2., 1., -1., 0.], [1., 2., 0., -1.]])
h = matrix([3., 3., 0., 0.])
sol = solvers.lp(c, G, h)
print(sol['x'])

[ 1.00e+00]
[ 1.00e+00]



## Model a Linear Program and Solve it using CVXOPT

Consider the problem of modeling a problem as linear programming problem.
A car company BNW has 4 manufacturing branches all over India. Number of labour units, materials, and pollution units are given as follows:

$$
\begin{array}{r|r}
& Labourers & Resources & Pollution \\ \hline
Factory 1 & 30 & 30 & 10 \\
Factory 2 & 30 & 60 & 5 \\
Factory 3 & 40 & 70 & 3 \\
Factory 4 & 20 & 30 & 11
\end{array}
$$

Let $x_1, x_2, x_3, x_4$ be the number of cars produced at these factories. The objective is to maximize the number of cars produced, given the following constraints:

<ol>
<li> Factory 3 must produce at least 4 cars
    <li> All other factories must produce at least 1 car
<li> The total labour hours available is 4000, and
<li> The number of material resources available is 16000
<li> Total units of pollution must not be more than 12000
</ol>


$\color{red}{Question:}$ Model the above problem as linear program.


$\color{blue}{Answer:}$

### Decision Variables:

Let:

- x1 = Number of cars produced by Factory 1
- x2 = Number of cars produced by Factory 2
- x3 = Number of cars produced by Factory 3
- x4 = Number of cars produced by Factory 4

### Objective Function:

We are to maximize the total number of cars produced:
Maximize Z = x1 + x2 + x3 + x4

### Constraints:

1. Factory 3 must produce at least 4 cars:
   x3 >= 4

2. All other factories must produce at least 1 car:
   x1 >= 1
   x2 >= 1
   x4 >= 1

3. The total labour hours available is 4000:
   30x1 + 30x2 + 40x3 + 20x4 <= 4000

4. The number of material resources available is 16000:
   30x1 + 60x2 + 70x3 + 30x4 <= 16000

5. Total units of pollution must not be more than 12000:
   10x1 + 5x2 + 3x3 + 11x4 <= 12000


$\color{red}{Question:}$ Write the constraint matrices and vectors.


$\color{blue}{Answer:}$

To model the Linear Program, we write it in standard form:

Minimize $$ c^T x $$
subject to 
$$ Ax \leq b $$
$$ x \geq 0 $$

### Coefficient Vector for Objective Function, $ c $

The objective is to maximize the total number of cars produced, which can be represented as minimizing the negative of the total number of cars produced.
$$ c = \begin{bmatrix} -1 \\ -1 \\ -1 \\ -1 \end{bmatrix} $$

### Constraint Coefficients Matrix, $ A $ and Right-Hand Side Vector, $ b $

- For the inequality constraints, the matrix $A$ and vector $b$ are defined as follows, incorporating all the given constraints including the minimum cars that need to be produced by each factory:
  $$
  A = \begin{bmatrix}
  30 & 30 & 40 & 20 \\
  30 & 60 & 70 & 30 \\
  10 & 5 & 3 & 11 \\
  -1 & 0 & 0 & 0 \\
  0 & -1 & 0 & 0 \\
  0 & 0 & -1 & 0 \\
  0 & 0 & 0 & -1 \\
  \end{bmatrix}
  $$

$$
b = \begin{bmatrix}
4000 \\
16000 \\
12000 \\
-1 \\
-1 \\
-4 \\
-1 \\
\end{bmatrix}
$$


$\color{red}{Question:}$ Write python code, and show solution.


In [13]:
from cvxopt import matrix, solvers

# objective function coefficients. -1 to maximize
c = matrix([-1., -1., -1., -1.])
G = matrix([  # constraint coefficients
    [30., 30., 10., -1., 0., 0., 0.],
    [30., 60., 5., 0., -1., 0., 0.],
    [40., 70., 3., 0., 0., -1., 0.],
    [20., 30., 11., 0., 0., 0., -1.]
])
# constraint RHS Gx <= h
h = matrix([4000., 16000., 12000., -1., -1., -4., -1.])
# no equality constraints
sol = solvers.lp(c, G, h)
print(sol['x'])

[ 1.00e+00]
[ 1.00e+00]
[ 4.00e+00]
[ 1.89e+02]



$\color{blue}{Answer:}$

The Python code snippet provided is using the `cvxopt` library to solve a linear programming problem. The steps to it have been explained in the comments.


Your solution should be:

[ 1.00e+00]
[ 1.00e+00]
[ 4.00e+00]
[ 1.89e+02]


# Quadratic Program

We can write a quadratic program as follows:

$$
\begin{align*}
\text{minimize} \quad &(1/2)x^T P x + q^T x \\
\text{subject to} \quad &Gx \leq h \\
&Ax = b
\end{align*}
$$


Consider the following quadratic program.

$$
\begin{align*}
\text{minimize} \quad &2x_1^2 + x_2^2 + x_1x_2 + x_1 + x_2 \\
\text{subject to} \quad &x_1 \geq 0 \\
&x_2 \geq 0 \\
&x_1 + 2x_2 = 1
\end{align*}
$$


$\color{red}{Question:}$ Identify $P, q, G, A, h, b$


$\color{red}{Answer:}$ Write your answer here.

1. **Matrix \( P \) (Quadratic Coefficients Matrix):**
   Represents the coefficients of the quadratic and mixed terms in the objective function.

   $$
   P = \begin{bmatrix} 4 & 1 \\ 1 & 2 \end{bmatrix}
   $$

2. **Vector \( q \) (Linear Coefficients Vector):**
   Represents the coefficients of the linear terms in the objective function.

   $$
   q = \begin{bmatrix} 1 \\ 1 \end{bmatrix}
   $$

3. **Matrix \( G \) (Inequality Constraint Coefficients Matrix):**
   Represents the coefficients of the inequality constraints.

   $$
   G = \begin{bmatrix} -1 & 0 \\ 0 & -1 \end{bmatrix}
   $$

4. **Vector \( h \) (Right-hand Side of Inequality Constraints Vector):**
   Represents the right-hand side of the inequality constraints.

   $$
   h = \begin{bmatrix} 0 \\ 0 \end{bmatrix}
   $$

5. **Matrix \( A \) (Equality Constraint Coefficients Matrix):**
   Represents the coefficients of the equality constraints.

   $$
   A = \begin{bmatrix} 1 & 2 \end{bmatrix}
   $$

6. **Vector \( b \) (Right-hand Side of Equality Constraints Vector):**
   Represents the right-hand side of the equality constraint.
   $$
   b = \begin{bmatrix} 1 \end{bmatrix}
   $$


$\color{red}{Question:}$ Complete the python code below by filling the matrices found above.


In [7]:
from cvxopt import matrix, solvers

Q = matrix([
    [4., 1.],
    [1., 2.]
])

p = matrix([1., 1.])

G = matrix([
    [-1., 0.],
    [0., -1.]
])

h = matrix([0., 0.])
A = matrix([1., 2.], (1, 2))
b = matrix([1.])

sol = solvers.qp(Q, p, G, h, A, b)
print(sol['x'])

     pcost       dcost       gap    pres   dres
 0:  7.2853e-01  2.8532e-01  3e+00  1e+00  2e+00
 1:  8.3896e-01  4.5876e-01  4e-01  6e-17  2e-16
 2:  7.6893e-01  7.4535e-01  2e-02  6e-17  8e-17
 3:  7.5086e-01  7.4997e-01  9e-04  6e-17  4e-17
 4:  7.5001e-01  7.5000e-01  9e-06  2e-16  2e-16
 5:  7.5000e-01  7.5000e-01  9e-08  1e-22  2e-16
Optimal solution found.
[ 1.74e-07]
[ 5.00e-01]



The solution should be:

[ 0]

[ 0.5]


When solving optimization problems numerically, solutions may not always be exact due to the inherent limitations of numerical precision in solvers and computer's floating-point representation. A solution like $1.74 \times 10^{-7}$ is practically equivalent to zero and is usually treated as such, especially when other values in the problem are significantly larger. In such cases, it is common to round such small values to zero, considering them as numerical artifacts.
